service.go

```diff
@@ -0,0 +1,541 @@
+package threads
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/ipfs/go-cid"
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+	"github.com/libp2p/go-libp2p-pubsub"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/textileio/go-textile-core/crypto"
+	"github.com/textileio/go-textile-core/crypto/asymmetric"
+	"github.com/textileio/go-textile-core/crypto/symmetric"
+	"github.com/textileio/go-textile-core/thread"
+	tserv "github.com/textileio/go-textile-core/threadservice"
+	"github.com/textileio/go-textile-threads/cbor"
+	pb "github.com/textileio/go-textile-threads/pb"
+	"github.com/textileio/go-textile-threads/util"
+	"google.golang.org/grpc"
+)
+
+const (
+	// reqTimeout is the duration to wait for a request to complete.
+	reqTimeout = time.Second * 5
+)
+
+// service implements the Threads RPC service.
+type service struct {
+	threads *threads
+	pubsub  *pubsub.PubSub
+}
+
+// newService creates a new threads network service.
+func newService(t *threads) (*service, error) {
+	ps, err := pubsub.NewGossipSub(
+		t.ctx,
+		t.host,
+		pubsub.WithMessageSigning(false),
+		pubsub.WithStrictSignatureVerification(false))
+	if err != nil {
+		return nil, err
+	}
+
+	s := &service{
+		threads: t,
+		pubsub:  ps,
+	}
+
+	for _, id := range t.Threads() {
+		go s.subscribe(id)
+	}
+
+	// @todo: ts.pubsub.RegisterTopicValidator()
+
+	return s, nil
+}
+
+// Push receives a record request.
+func (s *service) Push(ctx context.Context, req *pb.PushRequest) (*pb.PushReply, error) {
+	// Verify the request
+	reqpk, err := requestPubKey(req)
+	if err != nil {
+		return nil, err
+	}
+	err = verifyRequestSignature(req.Record, reqpk, req.Header.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	reply := &pb.PushReply{Ok: true}
+
+	// Unpack the record
+	var fkey []byte
+	if req.Header.FollowKey != nil {
+		fkey = req.Header.FollowKey
+	} else {
+		fkey = s.threads.FollowKey(req.ThreadID.ID, req.LogID.ID)
+	}
+	if fkey == nil {
+		return nil, fmt.Errorf("could not find follow key")
+	}
+
+	rec, err := recordFromProto(req.Record, fkey)
+	if err != nil {
+		return nil, err
+	}
+	x, err := s.threads.blocks.Blockstore().Has(rec.Cid())
+	if err != nil {
+		return nil, err
+	}
+	if x {
+		return reply, nil
+	}
+
+	// Check if this log already exists
+	logpk := s.threads.PubKey(req.ThreadID.ID, req.LogID.ID)
+	if logpk == nil {
+		var kid peer.ID
+		if req.Header.ReadKeyLogID != nil {
+			kid = req.Header.ReadKeyLogID.ID
```

Comment:
If this request indicates that the read key is from a different log, this means that this is an invite from a new peer that used the invitee's log (which we should have) to encrypt _their_ invite (which is effectively a join). I'm trying to avoid message types (like invite / join) in favor of just interpreting the context. Working so far.